File: src/weather_basis/schemas/public.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .base import StrictRecord, require


@dataclass(frozen=True, kw_only=True)
class ResultEnvelope(StrictRecord):
    release_id: str
    object_id: str
    result_type: str
    analysis_id: str
    source_artifact_ids: tuple[str, ...]
    data_vintage_id: str
    model_spec_ids: tuple[str, ...]
    scenario_set_id: str | None
    valuation_asof: str
    index_definition_id: str
    units: str
    currency: str | None
    payload: dict[str, Any] | None
    status: str
    reason_code: str | None
    evidence_reference: str

    def __post_init__(self):
        super().__post_init__()
        require(self.currency in (None, "USD"), "Unsupported public currency")
        require(bool(self.release_id and self.object_id and self.analysis_id), "Missing identity")
        require(bool(self.source_artifact_ids and self.evidence_reference), "Missing provenance")
        require(self.status in ("available", "unavailable", "partial"), "Invalid result status")
        if self.status != "available":
            require(bool(self.reason_code), "Unavailable/partial result requires reason")
        if self.status == "available":
            require(
                self.payload is not None and self.reason_code is None,
                "Available results require payload and no unavailable reason",
            )

File: apps/site/js/data/public.js
/** Strict decoder for the canonical V2 public ResultEnvelope boundary. */
const statuses = new Set(['available', 'unavailable', 'partial']);
export function assertResultEnvelope(value, { releaseId } = {}) {
  if (!value || typeof value !== 'object') throw new TypeError('Public object must be a JSON object.');
  const required = ['schema_version', 'release_id', 'object_id', 'result_type', 'analysis_id', 'source_artifact_ids', 'data_vintage_id', 'model_spec_ids', 'valuation_asof', 'index_definition_id', 'units', 'status', 'evidence_reference'];
  required.forEach((key) => { if (!(key in value)) throw new TypeError(`Public object is missing ${key}.`); });
  if (value.schema_version !== '2.0') throw new TypeError(`Unsupported public schema ${value.schema_version}.`);
  if (releaseId && value.release_id !== releaseId) throw new TypeError('Payload release does not match the committed scenario release.');
  if (!statuses.has(value.status)) throw new TypeError('Public object has an invalid availability status.');
  if (value.status === 'available' && (value.payload === null || value.reason_code !== null)) throw new TypeError('Available object has an invalid payload or reason.');
  if (value.status !== 'available' && !value.reason_code) throw new TypeError('Unavailable object is missing a reason code.');
  return value;
}

export function assertBootstrap(value, { releaseId } = {}) {
  if (!value || value.schema_version !== '2.0' || !value.release_id || !value.objects) throw new TypeError('Invalid V2 release bootstrap.');
  if (releaseId && value.release_id !== releaseId) throw new TypeError('Bootstrap release does not match the scenario link.');
  const objects = Array.isArray(value.objects) ? value.objects : Object.values(value.objects);
  objects.forEach((object) => {
    if (object?.schema_version) assertResultEnvelope(object, { releaseId: value.release_id });
    else if (!object?.path || typeof object.path !== 'string') throw new TypeError('Public object reference is missing its relative path.');
  });
  return value;
}

export function bootstrapObjects(bootstrap) {
  return Array.isArray(bootstrap.objects)
    ? bootstrap.objects
    : Object.entries(bootstrap.objects).map(([objectKey, value]) => ({ ...value, object_key: objectKey }));
}
